### PR TITLE
Use automatic memory management for OpenSSL objects

### PR DESCRIPTION
**Pull request by lbonn**
Freeing the right objects at the right time is tedious...

Based on #714.

### CODE REVIEW

**Comment by cajun-rat** [comment]
@lbonn I've just merged the PR this was built on. Can you rebase, please?

**Comment by lbonn** [comment]
Rebased

**Comment by cajun-rat** [comment]
What about something like:

~~~
class BioBuffer {
   BioBuffer() : bio_(BIO_new(BIO_s_mem())); {}
   explicit BioBuffer(const std::string&);
   // move assignment/ctor
   BIO* bio();
   std::string str() const;
 private:
  BIO* bio_;  
};
Crypto::parseP12(const BioBuffer&, ...);
~~~

File: src/utilities/crypto.cc [hunk 1]
```diff
@@ -275,72 +269,56 @@ bool Crypto::generateRSAKeyPair(KeyType key_type, std::string *public_key, std::
...
   if (!pubkey_sink) {
     return false;
   }
-  ret = PEM_write_bio_PUBKEY(pubkey_sink, pkey);
+  ret = PEM_write_bio_PUBKEY(pubkey_sink.get(), pkey.get());
   if (ret != 1) {
-    RSA_free(r);
-#if AKTUALIZR_OPENSSL_AFTER_11
-    BN_free(bne);
-#endif
```
Comment (cajun-rat) [inline src/utilities/crypto.cc]: Is this logic definitely not needed any more?  I can't see where it moved to. I think @OYTIS understands the reasons for it.

**Comment by lbonn** [comment]
@cajun-rat  yes actually it looks like it could work. I'll give it a go

File: src/utilities/crypto.cc (hunk 1 above)
Comment (cajun-rat) [reply src/utilities/crypto.cc]: Ah, OK. I'd missed that

**Pull request merged**